client/state/themes/actions/recommended-themes.js

```diff
@@ -26,13 +26,13 @@ export function receiveRecommendedThemes( themes, filter ) {
  * @param {string} filter A filter string for a theme query
  * @returns {Function} Action thunk
  */
-export function getRecommendedThemes( filter ) {
+export function getRecommendedThemes( filter, tier = '' ) {
 	return async ( dispatch ) => {
 		dispatch( { type: RECOMMENDED_THEMES_FETCH, filter } );
 		const query = {
 			search: '',
 			number: 50,
-			tier: '',
+			tier,
```

Comment:
This means the fetching takes the `tier` into account, but the reducer doesn't take the `tier` into account.
https://github.com/Automattic/wp-calypso/blob/031a60d1891e2c610d9e8048a241c7e523d8cef0/client/state/themes/reducer.js#L491

It will stash the response in the store using only the `filter`, which means when the selector gets called it will return whatever tier just happens to have been last fetched.

A fix for this would be to store the response in the redux using both the `filter` and `tier`. But since I'd [already planned on moving the theme fetching to `@automattic/design-picker`](https://github.com/Automattic/wp-calypso/blob/efcc23e2759523ec534ec148902c1a46fe3272d5/client/signup/steps/design-picker/index.jsx#L94) anyway, I thought the addition of premium themes would be a good opportunity to make that change. Making the code sharable with Gutenboarding rather than modifying the recommended themes reducer.

I'm making that change here: https://github.com/Automattic/wp-calypso/pull/59342

#59342 doesn't implement the actual feature that this PR does though i.e. it doesn't use `hasFeature` to decide whether to request premium themes or not. But I think that it'd be good to merge #59342 first so that we don't have to modify the Redux store.

Reply:
Oh cool, I don't notice that we use `action.filter` as the key of the themes. I think https://github.com/Automattic/wp-calypso/pull/59342 is a great refactor and I'm looking forward to merging it 👍